File: Sources/batchTests/Modules/Metric/metricTests.m
//
//  BatchTests
//
//  Copyright © Batch.com. All rights reserved.
//

#import <XCTest/XCTest.h>

@interface metricTests : XCTestCase

@end

@implementation metricTests

- (void)setUp {
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
}

- (void)testToDictionary {
    BACounter *counter = [[BACounter alloc] initWithName:@"counter_test_metric" andLabelNames:@"label", nil];
    [counter increment];

    NSDictionary *dict = [counter toDictionary];

    XCTAssertNotNil(dict);
    XCTAssertEqualObjects(dict[@"name"], @"counter_test_metric");
    XCTAssertEqualObjects(dict[@"type"], @"counter");
    XCTAssertNotNil(dict[@"values"]);
    XCTAssertTrue([dict[@"values"] isKindOfClass:[NSArray class]]);
    NSArray *values = dict[@"values"];
    XCTAssertEqual(values.count, 1);
    XCTAssertEqualObjects(values[0], @1.0);
}

- (void)testHasChildren {
    BACounter *counter = [[BACounter alloc] initWithName:@"counter_test_metric"];
    XCTAssertFalse([counter hasChanged]);
    [counter increment];
    XCTAssertTrue([counter hasChanged]);
}

- (void)testHasChanged {
    BACounter *counter = [[BACounter alloc] initWithName:@"counter_test_metric" andLabelNames:@"label", nil];
    XCTAssertFalse([counter hasChildren]);
    NSArray<NSString *> *labels = [[NSArray alloc] initWithObjects:@"label", nil];
    [[counter labels:labels] increment];
    XCTAssertTrue([counter hasChildren]);
}

- (void)testCopy {
    // Non regression test for sc-45635, where copyWithZone was not implemented correctly and
    // returned BAMetric instances when copying one of its subclasses
    BAMetric *metric = [[BAMetric alloc] initWithName:@"counter_test_metric" andLabelNames:@"label", nil];
    BACounter *counter = [[BACounter alloc] initWithName:@"counter_test_metric" andLabelNames:@"label", nil];
    BAObservation *observation = [[BAObservation alloc] initWithName:@"counter_test_metric"
                                                       andLabelNames:@"label", nil];

    XCTAssertEqual([[metric copy] class], BAMetric.class);
    XCTAssertEqual([[counter copy] class], BACounter.class);
    XCTAssertEqual([[observation copy] class], BAObservation.class);
}

@end

File: Sources/batchTests/Modules/Metric/metricTests.swift
//
//  metricTests.swift
//  batchTests
//
//  Copyright © Batch.com. All rights reserved.
//

import Foundation
import Testing

@testable import Batch

struct MetricTests {
    @Test func testToDictionary() throws {
        let counter = BACounter(name: "counter_test_metric", andLabelNamesList: ["label"])
        counter.increment()

        let dict = counter.toDictionary() as NSDictionary

        #expect(dict["name"] as? String == "counter_test_metric")
        #expect(dict["type"] as? String == "counter")

        let values = try #require(dict["values"] as? [NSNumber])
        #expect(values.count == 1)
        #expect(values[0].doubleValue == 1.0)
    }

    @Test func testHasChanged() {
        let counter = BACounter(name: "counter_test_metric")
        #expect(counter.hasChanged() == false)
        counter.increment()
        #expect(counter.hasChanged())
    }

    @Test func testHasChildren() {
        let counter = BACounter(name: "counter_test_metric", andLabelNamesList: ["label"])
        #expect(counter.hasChildren() == false)
        let child = counter.labels(["label"]) as? BACounter
        child?.increment()
        #expect(counter.hasChildren())
    }

    @Test func testCopy() {
        // Non regression test, copyWithZone was not implemented correctly and
        // returned BAMetric instances when copying one of its subclasses
        let metric = BAMetric(name: "counter_test_metric", andLabelNamesList: ["label"])
        let counter = BACounter(name: "counter_test_metric", andLabelNamesList: ["label"])
        let observation = BAObservation(name: "counter_test_metric", andLabelNamesList: ["label"])

        #expect(type(of: metric.copy() as AnyObject) == BAMetric.self)
        #expect(type(of: counter.copy() as AnyObject) == BACounter.self)
        #expect(type(of: observation.copy() as AnyObject) == BAObservation.self)
    }

    @Test func testObservationChildInheritsStartTime() throws {
        let observation = BAObservation(name: "observation_test_metric", andLabelNamesList: ["type"])
        observation.startTimer()
        Thread.sleep(forTimeInterval: 0.01)
        let child = observation.labels(["image"]) as? BAObservation
        child?.observeDuration()

        let values = try #require(child?.values)
        let firstValue = try #require(values.firstObject as? NSNumber)
        #expect(firstValue.doubleValue < 10.0)
    }
}
